src/Control/Effect/Carrier.hs

```diff
@@ -34,46 +31,109 @@ class HFunctor h where
-  eff :: sig m (m a) -> m a
+  eff :: sig m a -> m a
 
 -- | Apply a handler specified as a natural transformation to both higher-order and continuation positions within an 'HFunctor'.
-handlePure :: HFunctor sig => (forall x . f x -> g x) -> sig f (f a) -> sig g (g a)
-handlePure handler = hmap handler . fmap' handler
+handlePure :: (HFunctor sig, Functor f, Functor g) => (forall x . f x -> g x) -> sig f a -> sig g a
+handlePure = hmap
 {-# INLINE handlePure #-}
+{-# DEPRECATED handlePure "handlePure has been subsumed by hmap." #-}
 
 -- | Thread a 'Coercible' carrier through an 'HFunctor'.
 --
 --   This is applicable whenever @f@ is 'Coercible' to @g@, e.g. simple @newtype@s.
-handleCoercible :: (HFunctor sig, Coercible f g) => sig f (f a) -> sig g (g a)
-handleCoercible = handlePure coerce
+handleCoercible :: (HFunctor sig, Functor f, Functor g, Coercible f g) => sig f a -> sig g a
+handleCoercible = hmap coerce
 {-# INLINE handleCoercible #-}
+
+
+-- | Generic implementation of 'HFunctor'.
+class GHFunctor m m' rep rep' where
+  -- | Generic implementation of 'hmap'.
+  ghmap :: (Functor m, Functor m') => (forall x . m x -> m' x) -> (rep a -> rep' a)
+
+instance GHFunctor m m' rep rep' => GHFunctor m m' (M1 i c rep) (M1 i c rep') where
+  ghmap f = M1 . ghmap f . unM1
+
+instance (GHFunctor m m' l l', GHFunctor m m' r r') => GHFunctor m m' (l :+: r) (l' :+: r') where
+  ghmap f (L1 l) = L1 (ghmap f l)
+  ghmap f (R1 r) = R1 (ghmap f r)
+
+instance (GHFunctor m m' l l', GHFunctor m m' r r') => GHFunctor m m' (l :*: r) (l' :*: r') where
+  ghmap f (l :*: r) = ghmap f l :*: ghmap f r
+
+instance GHFunctor m m' V1 V1 where
+  ghmap _ v = case v of {}
+
+instance GHFunctor m m' U1 U1 where
+  ghmap _ = id
+
+instance GHFunctor m m' (K1 R c) (K1 R c) where
+  ghmap _ = coerce
+
+instance (Functor f, GHFunctor m m' g g') => GHFunctor m m' (f :.: g) (f :.: g') where
+  ghmap f = Comp1 . fmap (ghmap f) . unComp1
+
+instance GHFunctor m m' (Rec1 m) (Rec1 m') where
+  ghmap f = Rec1 . f . unRec1
+
+instance HFunctor f => GHFunctor m m' (Rec1 (f m)) (Rec1 (f m')) where
+  ghmap f = Rec1 . hmap f . unRec1
+
+
+-- | Generic implementation of 'Effect'.
+class GEffect m m' rep rep' where
+  -- | Generic implementation of 'handle'.
+  ghandle :: (Functor f, Monad m, Monad m')
+          => f ()
+          -> (forall x . f (m x) -> m' (f x))
+          -> rep a
+          -> rep' (f a)
+
+instance GEffect m m' rep rep' => GEffect m m' (M1 i c rep) (M1 i c rep') where
+  ghandle state handler = M1 . ghandle state handler . unM1
+
+instance (GEffect m m' l l', GEffect m m' r r') => GEffect m m' (l :+: r) (l' :+: r') where
+  ghandle state handler (L1 l) = L1 (ghandle state handler l)
+  ghandle state handler (R1 r) = R1 (ghandle state handler r)
+
+instance (GEffect m m' l l', GEffect m m' r r') => GEffect m m' (l :*: r) (l' :*: r') where
```

Comment:
This does not implement left-to-right threading of state through multiple higher-order positions, but that really only occurs when you’ve got existentials, which can’t use the generic derivations anyway.